src/lib.rs

```diff
@@ -150,110 +213,260 @@ impl<'object, Endian> DebugInfo<'object, Endian>
+                let directory = directory.to_string_lossy();
+                if !directory.starts_with("/") {
+                    if let Some(comp_dir) = unit.comp_dir() {
+                        path.push(&*comp_dir.to_string_lossy());
                     }
-                    break;
                 }
-                if row.end_sequence() {
-                    current = None;
+                path.push(&*directory);
+            }
+            path.push(&*file.path_name().to_string_lossy());
+
+            let line = row.line().unwrap();
+            if unit.programs.is_empty() {
+                return Some((path, line, None));
+            }
+
+            // The unit also has programs, so let's look for the function wrapping this address.
+            let mut func: Option<(&Program, &gimli::Range, u64)> = None;
+            for p in &unit.programs {
+                if !p.contains_address(addr) {
+                    continue;
+                }
+
+                // This program covers the given address -- calculate how well it matches
+                let (range, dist) = p.ranges
+                    .iter()
+                    .filter(|range| addr >= range.begin && addr < range.end)
+                    .map(|range| (range, addr - range.begin))
+                    .min_by_key(|&(_, dist)| dist)
+                    .unwrap();
+
+                if let Some((prev, prange, pdist)) = func.take() {
+                    // are we a better match?
+                    func = if dist == pdist {
+                        // we're equally good -- are we tighter?
+                        if range.end < prange.end {
+                            // yes!
+                            Some((p, range, dist))
+                        } else if range.end == prange.end {
+                            println!("found two equivalent ranges for {:x}: {} and {}",
+                                     addr,
+                                     prev.name.to_string_lossy(),
+                                     p.name.to_string_lossy());
+                            Some((p, range, dist))
+                        } else {
+                            // no
+                            Some((prev, prange, pdist))
+                        }
+                    } else if dist < pdist {
+                        // we're just better
+                        Some((p, range, dist))
+                    } else {
+                        // current is better
+                        Some((prev, prange, pdist))
+                    };
                 } else {
-                    current = Some(row.clone());
+                    func = Some((p, range, dist));
                 }
+
+                // Unfortunately, we're not done even if we've already found loc, because there
+                // *may* be another subprogram (specifically, one that has been inlined) that
+                // matches us better. We need to keep going until we find the best one.
```

Comment:
It would be nice to (optionally) return the whole set of virtual frames (one physical frame and all the frames inlined into it) instead of only the innermost inlined frame.

A profiler that is symbolicating a stack of physical frames would certainly like to have this information...

For another PR, either way.